src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
-    core::smart_refctd_ptr<const CIntrospectionData>* const introspections, const std::span<const ICPUShader::SSpecInfo> infos
-)
+NBL_API2 core::smart_refctd_ptr<ICPUPipelineLayout> CSPIRVIntrospector::CPipelineIntrospectionData::createApproximatePipelineLayoutFromIntrospection(
+    core::smart_refctd_ptr<const CStageIntrospectionData>& introspection)
 {
-    core::smart_refctd_ptr<ICPUDescriptorSetLayout> dsLayout[ICPUPipelineLayout::DESCRIPTOR_SET_COUNT];
-    for (uint32_t i = 0u; i<ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
-        dsLayout[i] = createApproximateDescriptorSetLayoutFromIntrospection_impl(i,introspections,infos);
-
-    auto pcRanges = createPushConstantRangesFromIntrospection_impl(introspections,infos);
+    const auto pcRanges = createPushConstantRangesFromIntrospection(introspection);
+    const std::span<const asset::SPushConstantRange> pcRangesSpan = {
+        pcRanges ? pcRanges->begin() : nullptr,
+        pcRanges ? pcRanges->end() : nullptr
+    };
 
     return core::make_smart_refctd_ptr<ICPUPipelineLayout>(
-        std::span<const SPushConstantRange>(pcRanges ? pcRanges->data():nullptr,pcRanges ? pcRanges->size():0),
-        std::move(dsLayout[0]), std::move(dsLayout[1]), std::move(dsLayout[2]), std::move(dsLayout[3])
+        pcRangesSpan,
+        createApproximateDescriptorSetLayoutFromIntrospection(0),
+        createApproximateDescriptorSetLayoutFromIntrospection(1),
+        createApproximateDescriptorSetLayoutFromIntrospection(2),
+        createApproximateDescriptorSetLayoutFromIntrospection(3)
     );
 }
 
-static E_FORMAT glslType2E_FORMAT(E_GLSL_VAR_TYPE _t, uint32_t _e)
+CSPIRVIntrospector::CStageIntrospectionData::SDescriptorVarInfo<true>* CSPIRVIntrospector::CStageIntrospectionData::addResource(
+    const spirv_cross::Compiler& comp,
+    const spirv_cross::Resource& r,
+    IDescriptor::E_TYPE restype
+)
 {
-    static const E_FORMAT retval[6][4]
-    {
-        {EF_R64_UINT, EF_R64G64_UINT, EF_R64G64B64_UINT, EF_R64G64B64A64_UINT},
-        {EF_R64_SINT, EF_R64G64_SINT, EF_R64G64B64_SINT, EF_R64G64B64A64_SINT},
-        {EF_R32_UINT, EF_R32G32_UINT, EF_R32G32B32_UINT, EF_R32G32B32A32_UINT},
-        {EF_R32_SINT, EF_R32G32_SINT, EF_R32G32B32_SINT, EF_R32G32B32A32_SINT},
-        {EF_R64_SFLOAT, EF_R64G64_SFLOAT, EF_R64G64B64_SFLOAT, EF_R64G64B64A64_SFLOAT},
-        {EF_R32_SFLOAT, EF_R32G32_SFLOAT, EF_R32G32B32_SFLOAT, EF_R32G32B32A32_SFLOAT}
+    const uint32_t descSet = comp.get_decoration(r.id,spv::DecorationDescriptorSet);
+    if (descSet> DESCRIPTOR_SET_COUNT)
+        return nullptr; // TODO: log fail
+
+    auto a = comp.get_name(r.id);
+    const spirv_cross::SPIRType& type = comp.get_type(r.type_id);
+    const auto arrDim = type.array.size();
+    // assuming only 1D arrays because i don't know how desc set layout binding is constructed when it's let's say 2D array (e.g. uniform sampler2D smplr[4][5]; is it even legal?)
```

Comment:
definitely nnot legal in HLSL2021, just checked with all resource types except UBO, Input and Output

Reply:
can leave a comment about section 15.5.3 of Vulkan spec saying descirptor arrays have to be 1D